politeiawww/addresswatcher.go

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) 2019 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/decred/dcrd/dcrutil"
+	client "github.com/decred/dcrdata/pubsub/psclient"
+	pstypes "github.com/decred/dcrdata/pubsub/types"
+	cms "github.com/decred/politeia/politeiawww/api/cms/v1"
+	database "github.com/decred/politeia/politeiawww/cmsdatabase"
+	"github.com/decred/politeia/util"
+)
+
+func (p *politeiawww) addWatchAddress(address string) error {
+	address = "address:" + address
+	if subd, _ := strInSlice(p.currentSubs, address); subd {
+		log.Infof("Already subscribed to %s.", address)
+		return nil
+	}
+	p.currentSubs = append(p.currentSubs, address)
+	resp, err := p.wsClient.Subscribe(address)
+	if err != nil {
+		return fmt.Errorf("Failed to subscribe: %v", err)
+	}
+	log.Infof(resp.Data)
+	return nil
+}
+
+func (p *politeiawww) removeWatchAddress(address string) error {
+	address = "address:" + address
+	subd, i := strInSlice(p.currentSubs, address)
+	if !subd {
+		log.Infof("Not subscribed to %s.", address)
+		return nil
+	}
+	p.currentSubs = append(p.currentSubs[:i], p.currentSubs[i+1:]...)
+	resp, err := p.wsClient.Unsubscribe(address)
+	if err != nil {
+		return fmt.Errorf("Failed to unsubscribe: %v", err)
+	}
+	log.Infof(resp.Data)
+	return nil
+}
+
+func (p *politeiawww) addPing() error {
+	resp, err := p.wsClient.Subscribe("ping")
+	if err != nil {
+		return fmt.Errorf("Failed to subscribe: %v", err)
+	}
+	log.Infof(resp.Data)
+	return nil
+}
+
+func (p *politeiawww) setupWatcher() error {
+	// Create the websocket connection.
+	wsURL, err := util.BlockExplorerURLForSubscriptions(activeNetParams.Params)
+	if err != nil {
+		return err
+	}
+	log.Infof("connecting to ws at: %v", wsURL)
+
+	// XXX How to handle the ws close?
+	//defer ws.Close()
+	ctx, cancel := context.WithCancel(context.Background())
+	opts := client.Opts{
+		ReadTimeout:  3 * time.Second,
+		WriteTimeout: 3 * time.Second,
+	}
+	p.cancelContext = cancel
+	p.wsClient, err = client.New(wsURL, ctx, &opts)
+	if err != nil {
+		log.Errorf("Failed to connect to %s: %v", wsURL, err)
+		return err
+	}
+	go func() {
+		for {
+			msg := <-p.wsClient.Receive()
+			if msg == nil {
+				log.Errorf("ReceiveMsg failed")
```

Comment:
A few thousand of these `ReceiveMsg failed` are being logged when you kill the politeiawww process.

```
2019-05-31 16:18:53.095 [ERR] PWWW: ReceiveMsg failed
2019-05-31 16:18:53.095 [ERR] PWWW: ReceiveMsg failed
2019-05-31 16:18:53.095 [ERR] PWWW: ReceiveMsg failed
2019-05-31 16:18:53.095 [ERR] PWWW: ReceiveMsg failed
2019-05-31 16:18:53.095 [ERR] PWWW: ReceiveMsg failed
...
```

Reply:
Will investigate this, any ideas @chappjc ?  

Maybe I'm not canceling the context properly?  Only seems to occur on app shutdown.

Reply:
It's just that there's a `continue` statement here so is keeps trying to receive from the closed channel until the process terminates.

Probably should check for closed channel status and break instead of continue.  i.e.:

```go
msg, ok := <-p.wsClient.Receive()
if !ok {
	break // or return
}
if msg == nil {
	log.Errorf("ReceiveMsg failed")
	contine
}
```